ee/vellum_ee/workflows/display/tests/workflow_serialization/generic_nodes/test_attributes_serialization.py

```diff
@@ -670,3 +670,51 @@ class GenericNodeWithPydantic(BaseNode):
     assert any(
         entry["key"] == "node_ref" and entry["value"]["type"] == "NODE_OUTPUT" for entry in attr_value["entries"]
     )
+
+
+def test_serialize_node__dataclass_with_preserved_type_metadata(serialize_node):
+    """Test that dataclass type metadata is preserved during serialization."""
+
+    @dataclass
+    class CustomOutputType:
+        name: str
+        value: int
+
+    class NodeWithCustomDataclass(BaseNode):
+        attr = CustomOutputType(name="test", value=42)
+
+    serialized_node = serialize_node(node_class=NodeWithCustomDataclass)
+
+    attr_value = serialized_node["attributes"][0]["value"]
+    assert attr_value["type"] == "DICTIONARY_REFERENCE"
+
+    assert "dataclass_type" in attr_value
+    assert attr_value["dataclass_type"]["name"] == "CustomOutputType"
+    expected_module = (
+        "vellum_ee.workflows.display.tests.workflow_serialization.generic_nodes.test_attributes_serialization"
+    )
+    assert attr_value["dataclass_type"]["module"] == expected_module
+
+
+def test_serialize_node__dataclass_output_with_preserved_type_metadata(serialize_node):
+    """Test that dataclass outputs preserve type metadata during serialization."""
+
+    @dataclass
+    class CustomDataclassOutput:
+        result: str
+        count: int
+
+    class NodeWithDataclassOutput(BaseNode):
+        class Outputs(BaseNode.Outputs):
+            custom_output: CustomDataclassOutput
+
+    serialized_node = serialize_node(node_class=NodeWithDataclassOutput)
+
+    output = serialized_node["outputs"][0]
+    assert output["name"] == "custom_output"
+    assert output["type"] == "JSON"
+
+    if output["value"] is not None:
+        assert output["value"]["type"] == "DICTIONARY_REFERENCE"
+        assert "dataclass_type" in output["value"]
+        assert output["value"]["dataclass_type"]["name"] == "CustomDataclassOutput"
```

Comment:
Here's a previous TDD that we aligned on for what shape we should use: https://www.notion.so/vellum-ai/TDD-Custom-Vellum-Variable-Types-1619035c57bd80d09165f4876c52e810?source=copy_link#1619035c57bd818b8a03dc99c0b57032

Take a look at that section and improve the expected serialization logic here consistent with that doc. If you can't reference the Notion doc for whatever reason, ask me for more information. You should be able to log in using the credentials stored in the `~/.envrc`